### DISCUSSION TITLE
is it possible to find absolutely all cycles in a graph?

### DISCUSSION THREAD

**Discussion by a1nsworth** [networkx/networkx#7530] [Algorithms]
I need find absolutely all cycles(not only simple cycle) in graph with star concrete point value.

**Comment by satemochi** [comment]
I have a solution which is theoretically possible but not practical.

To begin, I would like to reach a common understanding. We confirm that the problem for finding a simple cycle can reduce to a somewhat rich problem:

Let $G = (V, E)$ be the input and $C = \\{~ c_i$  is the cycle graph with $i$ vertices  $|~~ i \in [|V|]\\}$. Then, find a (graph) _monomorphism_ from $c$ to $G$ for any $c \in C$. 

For more information on monomorphism, see [Stack Overflow](https://stackoverflow.com/questions/459799/whats-the-difference-between-subgraph-isomorphism-and-subgraph-monomorphism). Briefly, a map $f: V(c) \to V$ is a monomorphism from $c$ to $G$ then $f$ satisfies that if $\\{u, v\\} \in E(c)$ then $\\{f(u), f(v)\\} \in E,$ and $f$ is injection.

Your specification is to relax simplicity. This is equivalent to finding a _homomorphism_. That is, to remove injection from the definition of monomorphism. Note, however, that $C$ cannot be bounded since there can be infinitely many cycles in general.

Thus, the algorithm needs to solve enumerating homomorphisms to $G$ for all cycle graphs, and rejecting duplications; although not directly provided in NetworkX (as I know), homomorphism enumeration can be implemented by hacking the VF2 algorithm or `GraphMatcher`. Checking for duplication of cycles can also be solved by fundamental observation, or recognizing equivalence classes of permutation (sub)groups.

However, it is not acceptable to be fatally inefficient in both complexities for time and space, is it?

**Reply by dschult** [reply]
We have monomorphism iterators for vf2:  [subgraph_monomorphisms_iter](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.isomorphism.GraphMatcher.subgraph_monomorphisms_iter.html#networkx.algorithms.isomorphism.GraphMatcher.subgraph_monomorphisms_iter) is the method name.

**Reply by satemochi** [reply]
My hacking code for homomorphism is [this](https://github.com/satemochi/saaaaah/blob/master/geometric_misc/isomorph/gi/gi.py) (I wrote 5 years ago but probably executable).

If interest please see,
- `matcher` (vertual interface), and
- `extra_homo` (concrete implementation).
